cloudsmith/resource_service.go

```diff
@@ -152,29 +153,32 @@ func resourceServiceRead(ctx context.Context, d *schema.ResourceData, m interfac
 	// they'll need to recreate the resource if they want to pull the new key
 	// into Terraform. This can be accomplished by tainting.
 	var diags diag.Diagnostics
-	existingKey := requiredString(d, "key")
-	if existingKey == importSentinel {
-		diags = append(diags, diag.Diagnostic{
-			Severity: diag.Warning,
-			Summary:  "API key unavailable for imported services",
-			Detail: "API keys are only available via the Cloudsmith API at the time a service " +
-				"is created, and therefore it is not possible to retrieve the current API key for " +
-				"a service which has been imported. If the API key value is needed within Terraform" +
-				"then the resource can be tainted post-import to recreate it and store the key.",
-			AttributePath: cty.Path{cty.GetAttrStep{Name: "key"}},
-		})
-	} else {
-		existingLastFour := existingKey[len(existingKey)-4:]
-		newLastFour := service.GetKey()[len(service.GetKey())-4:]
-		if existingLastFour != newLastFour {
+	if requiredBool(d, "return_api_key") {
+		existingKey := requiredString(d, "key")
+		if existingKey == importSentinel {
 			diags = append(diags, diag.Diagnostic{
 				Severity: diag.Warning,
-				Summary:  "API key has changed",
-				Detail: "API key for this service has changed outside of Terraform. If this " +
-					"key is used within Terraform the resource must be tainted or otherwise " +
-					"recreated to retrieve the new value.",
+				Summary:  "API key unavailable for imported services",
+				Detail: "API keys are only available via the Cloudsmith API at the time a service " +
+					"is created, and therefore it is not possible to retrieve the current API key for " +
+					"a service which has been imported. If the API key value is needed within Terraform" +
+					"then the resource can be tainted post-import to recreate it and store the key.",
 				AttributePath: cty.Path{cty.GetAttrStep{Name: "key"}},
 			})
+		} else {
+			d.Set("key", "disabled")
```

Comment:
Question: Is the `key` output being set to `disabled` here to avoid the warning from appearing again?

I think this can cause an issue to users if they have other resources that depend on this field.

Reply:
@estebangarcia - the `key` is only being set to `disabled` if the flag `show_api_key` is set to `false`. The schema is set to `true` by default to not disturb anyone currently using the `key` field.

Reply:
Got it. The else conditional is set for the `if existingKey == importSentinel` when `return_api_key` is `true`

Might have to move it to the outer conditional?

Reply:
Ah, good spot! 

Do you mean something like this instead?:

```
	var diags diag.Diagnostics
	if requiredBool(d, "store_api_key") {
		existingKey := requiredString(d, "key")
		if existingKey == importSentinel {
			diags = append(diags, diag.Diagnostic{
				Severity: diag.Warning,
				Summary:  "API key unavailable for imported services",
				Detail: "API keys are only available via the Cloudsmith API at the time a service " +
					"is created, and therefore it is not possible to retrieve the current API key for " +
					"a service which has been imported. If the API key value is needed within Terraform" +
					"then the resource can be tainted post-import to recreate it and store the key.",
				AttributePath: cty.Path{cty.GetAttrStep{Name: "key"}},
			})
		} else {
			existingLastFour := existingKey[len(existingKey)-lastFourChars:]
			newLastFour := service.GetKey()[len(service.GetKey())-lastFourChars:]
			if existingLastFour != newLastFour {
				diags = append(diags, diag.Diagnostic{
					Severity: diag.Warning,
					Summary:  "API key has changed",
					Detail: "API key for this service has changed outside of Terraform. If this " +
						"key is used within Terraform, the resource must be tainted or otherwise " +
						"recreated to retrieve the new value.",
					AttributePath: cty.Path{cty.GetAttrStep{Name: "key"}},
				})
			}
		}
	} else {
		d.Set("key", "**redacted**")
	}
```

Reply:
Yeah exactly. With the previous version of the code `d.Set("key", "**redacted**")` will only execute when `store_api_key=true` and `existingKey !`= importSentinel`

Reply:
awesome, thanks! I've updated and pushed the new code.